cupy/cuda/memory.pyx

```diff
@@ -702,6 +712,88 @@ cdef class SingleDeviceMemoryPool:
 
         self._append_to_free_list(chunk.size, chunk)
 
+    cpdef _realloc(self, size_t ptr, Py_ssize_t size):
+        """Reallocate the given area of memory to mitigate fragmentation."""
+        try:
+            new_memptr = self._allocator(size)
+        except runtime.CUDARuntimeError as e:
+            if e.status != runtime.errorMemoryAllocation:
+                raise
+            runtime.deviceSynchronize()
+            return None
+
+        runtime.memcpy(new_memptr.ptr, ptr, size,
+                       runtime.memcpyDeviceToDevice)
+        runtime.deviceSynchronize()
+        return new_memptr
+
+    cpdef _reset_memptr(self, Chunk chunk, Chunk new_chunk):
+        if chunk.ptr in self._in_use:
+            chunk = self._in_use.pop(chunk.ptr)
+            memptr_ref = self._in_use_memptr.pop(chunk.ptr)
+            self._in_use[new_chunk.ptr] = new_chunk
+            self._in_use_memptr[new_chunk.ptr] = memptr_ref
+
+            pmem = PooledMemory(new_chunk, self._weakref)
+            memptr = memptr_ref()
+            memptr.mem.ptr = 0  # to skip free
+            try:
+                memptr.reset(pmem, 0)
+            except AttributeError:
+                # memptr is removed from app layer in another thread
+                pass
+        else:
+            if self._remove_from_free_list(chunk.size, chunk):
+                self._append_to_free_list(new_chunk.size, new_chunk)
+
+    cpdef _realloc_all(self):
+        """Reallocate all memory of in-use to mitigate fragmentation."""
+        ptrs_of_size = {}  # pointers of head chunks
+        rlock.lock_fastrlock(self._in_use_lock, -1, True)
+        try:
+            for ptr, chunk in self._in_use.iteritems():
+                size = chunk.mem.size
+                if size not in ptrs_of_size:
+                    ptrs_of_size[size] = set()
+                head_ptr = chunk.mem.ptr
+                ptrs_of_size[size].add(head_ptr)
+        finally:
+            rlock.unlock_fastrlock(self._in_use_lock)
+
+        # reallocate from smaller blocks because we assume _realloc_all
+        # is called when little memory is left
+        for size in sorted(ptrs_of_size.keys()):
+            for ptr in sorted(ptrs_of_size[size]):
+                new_memptr = self._realloc(ptr, size)
+                if new_memptr is None:
+                    return  # give up
+                if new_memptr.ptr > ptr:
+                    # This is based on an assumption that logical address are
+                    # almost linearly mapped to physical address. If the
+                    # assumption is correct, you can make larger continuous
+                    # free space in the high address area by calling the
+                    # realloc() twice when new address > old address. Note that
+                    # there is no guarantee that the assumption is correct.
+                    new_memptr_2nd = self._realloc(ptr, size)
```

Comment:
This 2nd re-allocation should be tried after new chunk(s) is creased based on 1st re-allocation and original pointer is released. Or, you will not get expected results.

Reply:
okay, i will fix